internal/pkg/build/buildkit/client/client.go

```diff
@@ -0,0 +1,249 @@
+	bkdaemon "github.com/sylabs/singularity/v4/internal/pkg/build/buildkit/daemon"
+	"github.com/sylabs/singularity/v4/internal/pkg/client/ocisif"
+	"github.com/sylabs/singularity/v4/internal/pkg/remote/credential/ociauth"
+	"github.com/sylabs/singularity/v4/pkg/sylog"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	buildTag        = "tag"
+	bkDefaultSocket = "unix:///run/buildkit/buildkitd.sock"
+	bkLaunchTimeout = 30 * time.Second
+)
+
+type Opts struct {
+	// Optional Docker authentication config derived from interactive login or
+	// environment variables
+	AuthConf *ocitypes.DockerAuthConfig
+	// Optional user requested authentication file for writing/reading OCI
+	// registry credentials
+	ReqAuthFile string
+	// Variables passed to build procedure.
+	BuildVarArgs []string
+	// Variables file passed to build procedure.
+	BuildVarArgFile string
+}
+
+func Run(ctx context.Context, opts *Opts, dest, spec string) {
+	listenSocket := ensureBuildkitd(ctx)
+	if listenSocket == "" {
+		sylog.Fatalf("Failed to launch buildkitd daemon within specified timeout (%v).", bkLaunchTimeout)
+	}
+
+	tarFile, err := os.CreateTemp("", "singularity-buildkit-tar-")
+	if err != nil {
+		sylog.Fatalf("While trying to build tar image from dockerfile: %v", err)
+	}
+	defer tarFile.Close()
+	defer os.Remove(tarFile.Name())
+
+	if err := buildImage(ctx, opts, tarFile, listenSocket, spec, false); err != nil {
+		sylog.Fatalf("While building from dockerfile: %v", err)
+	}
+	sylog.Debugf("Saved OCI image as tar: %s", tarFile.Name())
+	tarFile.Close()
+
+	if _, err := ocisif.PullOCISIF(ctx, nil, dest, "oci-archive:"+tarFile.Name(), ocisif.PullOptions{}); err != nil {
+		sylog.Fatalf("While converting OCI tar image to OCI-SIF: %v", err)
+	}
+}
+
+// ensureBuildkitd checks if a buildkitd daemon is already running, and if not,
+// launches one. Once the server is ready, the value true will be sent over the
+// provided readyChan. Make sure this is a buffered channel with sufficient room
+// to avoid deadlocks.
+func ensureBuildkitd(ctx context.Context) string {
+	if isBuildkitdRunning(ctx) {
+		sylog.Infof("Found buildkitd already running at %q; will use that daemon.", bkDefaultSocket)
+		return bkDefaultSocket
+	}
+
+	sylog.Infof("Did not find usable running buildkitd daemon; spawning our own.")
+	socketChan := make(chan string, 1)
+	go func() {
+		if err := bkdaemon.Run(ctx, socketChan); err != nil {
+			sylog.Fatalf("buildkitd returned error: %v", err)
+		}
+	}()
+	go func() {
+		time.Sleep(bkLaunchTimeout)
+		socketChan <- ""
+	}()
+
+	return <-socketChan
+}
+
+// isBuildkitdRunning tries to determine whether there's already an instance of buildkitd running.
+func isBuildkitdRunning(ctx context.Context) bool {
+	c, err := client.New(ctx, bkDefaultSocket, client.WithFailFast())
+	if err != nil {
+		return false
+	}
+	defer c.Close()
+
+	cc := c.ControlClient()
+	ir := moby_buildkit_v1.InfoRequest{}
+	_, err = cc.Info(ctx, &ir)
```

Comment:
Good question. The `InfoResponse` struct (the first return value of `cc.Info(...)` has the buildkit version in it... but what would our cutoff be?

Reply:
We _could_ set the cutoff at v0.12.3 just because that's what we developed against, and we can always launch an ephemeral buildkit of our own if the existing one is older... 🤷‍♂️

Reply:
That seems sensible to me.